### ISSUE TITLE
With unigram algorithm, constant piece at end of each sentences does not become a token

### ISSUE THREAD

**Issue by jogardi** [google/sentencepiece#1047]
Hi thanks for your great work on this. I noticed a subtle issue when playing with synthetic examples. 

The bpe algorithm works as expected but the unigram algorithm does not make this constant piece a token in the vocabulary.
I generate synthetic data where each sentence is a random string followed by a constant piece.
```
constant_piece = 'helloWorld'
def rand_str(n=10):
    return ''.join(
        np.random.choice(list('bcegijklmnoqruvwxyz'), n)
    )

data = [rand_str() + constant_piece for _ in range(1000)]
model = io.BytesIO()
spm.SentencePieceTrainer.train(
      sentence_iterator=iter(data), model_writer=model, 
    vocab_size=1000,
    minloglevel=5, 
)
sp = spm.SentencePieceProcessor(model_proto=model.getvalue())

ex = data[20]
print([
    sp.IdToPiece(x)
    for x in sp.encode(ex, emit_unk_piece=True)
])
```
outputs: ['▁uy', 'vx', 'yf', 'p', 'gmn', 'he', 'llo', 'W', 'or', 'ld']

It mostly just gets random tokens. I think it gets 'he', 'llo', 'or' and 'ld' not because it noticed the repeating pattern but just by coincidently seeing it in the random strings. If I change constant_piece to '123456' then i get no tokens for the repeating pattern and only tokens for the random string: `['▁', 'gll', 'imq', 'xc', 'df', '1', '2', '3', '4', '5', '6']`

This specifically because the constant_piece at the end. If I change data so that constant_piece is at the beginning of each sentence: `data = [constant_piece + rand_str() for _ in range(1000)]` then i get the expected result `['▁123456', 'uzb', 'ek', 'hoe', 'wr']`. 

**TLDR;**
Unexpected result under the following conditions:
- same string at end of each sentence in the training data
- using unigram algorithm

**Comment by DSS3113** [comment]
Hi, so I investigated this issue and I believe this is a limitation in how the unigram trainer extracts seed sentencepieces. The culprit is in `src/unigram_model_trainer.cc` around line 282. <br>
The unigram trainer concatenates all sentences into a single array with boundary markers (`0x0000`) placed after each sentence, then uses a suffix array to find frequent substrings. Any substring containing a boundary marker gets filtered out. When your constant piece is at the end, all 1000 occurrences are followed by the same boundary marker, so the suffix array identifies `helloWorld\0` as the common pattern, which then gets rejected. When it is at the beginning, the constant piece is followed by different random characters each time, so the suffix array correctly identifies just `helloWorld` as the longest common prefix (no boundary contamination). <br>
You could try adding the argument `user_defined_symbols=['helloWorld']` in your `train` method or you could add trailing whitespace/content after the constant piece so the boundary marker isn't immediately adjacent. <br>

**Comment by taku910** [comment]
Fixed in the head. Thank you.

https://github.com/google/sentencepiece/commit/3a6c0a4b790f3953efa1201ab30e8d8f0fe6f956
